lib/store.js

```diff
@@ -179,7 +335,144 @@ class Store {
     if (this.keyWhiteList && this.keyWhiteList.indexOf(key) < 0) {
       return Promise.reject(new Error('Saving key is not permitted'));
     }
-    return this._driver.setItem(key, value, callback);
+    return this._setItemWithRetry(key, value, callback);
+  }
+
+  _setItemWithRetry(key, value, callback, retryCount = 0) {
+    return this._driver.setItem(key, value).then(function () {
+      if (callback) {
+        callback(null);
+      }
+      return Promise.resolve();
+    }, function (error) {
+      if (retryCount >= this._maxRetryCount) {
+        const e = new Error('exceeded max retry count');
+        if (callback) {
+          callback(e);
+        }
+        return Promise.reject(e);
+      }
+      return this._performRecovery().then(function () {
+        return this._setItemWithRetry(key, value, callback, retryCount + 1);
```

Comment:
Convert to use `for..` loop instead of recursive?

It will be easier to read and reason about. Recursive with promise nest is killing the readability.